File: expo/app/src/flow/use-smile-id-sample-flow-tokens.ts
import {
  smileIDSampleBase64UrlEncode,
  smileIDSampleEnvironmentBaseUrl,
  type UseSmileIDSampleEnvironment,
  type UseSmileIDSampleSimulatedBindings,
  type UseSmileIDSampleSimulatedSpan,
} from '@smileid/sample-ui';

/// Structurally valid unsigned JWTs — fixtures the scenarios demand, never credentials.
export const smileIDSampleFlowToken = ({
  expired,
  nowMillis,
}: {
  readonly expired: boolean;
  readonly nowMillis: number;
}): string => {
  const exp = Math.floor(nowMillis / millisPerSecond) + (expired ? -validitySeconds : validitySeconds);
  return [header, `{"exp":${exp}}`, signature].map(smileIDSampleBase64UrlEncode).join('.');
};

/// What `badRefresh` hands back, so the failure path has something unusable to reject.
export const smileIDSampleMalformedToken = (): string => 'sample-not-a-jwt';

/// A simulated scan's unsigned token over the chosen span and bindings, with nonsense PII.
export const smileIDSampleSimulatedToken = ({
  span,
  bindings,
  environment,
  nowMillis,
}: {
  readonly span: UseSmileIDSampleSimulatedSpan;
  readonly bindings: UseSmileIDSampleSimulatedBindings;
  readonly environment: UseSmileIDSampleEnvironment;
  readonly nowMillis: number;
}): string => {
  const nowSeconds = Math.floor(nowMillis / millisPerSecond);
  const spanSeconds = span.spanMillis / millisPerSecond;
  // An ended span is minted in the past, the only way to reach the expiry gate.
  const issuedAt = span.ended ? nowSeconds - spanSeconds - endedLagSeconds : nowSeconds;
  const claims = [
    `"iat":${issuedAt}`,
    `"exp":${issuedAt + spanSeconds}`,
    // With a real claim's path, so the host match is exercised.
    `"api_url":"${smileIDSampleEnvironmentBaseUrl(environment)}${apiPath}"`,
  ];
  if (bindings.consent || bindings.userDetails) claims.push(payloadClaim(bindings, issuedAt));
  return [header, `{${claims.join(',')}}`, signature].map(smileIDSampleBase64UrlEncode).join('.');
};

const payloadClaim = (bindings: UseSmileIDSampleSimulatedBindings, issuedAtSeconds: number): string => {
  const fields: string[] = [];
  if (bindings.userDetails) {
    for (const field of vaultedFields) fields.push(`"${field}":"vault_${field}"`);
    // Plaintext on a Portal token too.
    fields.push('"country":"KE"', '"id_type":"NATIONAL_ID"');
  }
  if (bindings.consent) fields.push(consentClaim(issuedAtSeconds));
  return `"payload":{${fields.join(',')}}`;
};

/// All four subfields: a partial binding is a build error.
const consentClaim = (issuedAtSeconds: number): string => {
  const grantedAt = new Date(issuedAtSeconds * millisPerSecond).toISOString().replace(/\.\d{3}Z$/, 'Z');
  return (
    `"consent":{"granted":true,"granted_at":"${grantedAt}",` +
    `"notice_language":"en","notice_privacy_policy_url":"${privacyPolicyUrl}"}`
  );
};

const header = '{"alg":"none","typ":"JWT"}';
const signature = 'sample-signature';
const millisPerSecond = 1000;
const validitySeconds = 3600;
const endedLagSeconds = 60;
const apiPath = 'v3';
const privacyPolicyUrl = 'https://smile.id/privacy-policy';
const vaultedFields = ['given_names', 'last_name', 'email', 'phone_number', 'id_number'];

File: flutter/app/lib/src/flow/use_smileid_sample_flow_tokens.dart
import 'dart:convert';

import 'package:sample_ui/sample_ui.dart';

/// Structurally valid unsigned JWTs — fixtures the scenarios demand, never credentials.
abstract final class UseSmileIDSampleFlowTokens {
  /// A token whose `exp` is an hour away, or an hour past when [expired].
  static String token({required bool expired, required int nowMillis}) {
    final int exp =
        nowMillis ~/ _millisPerSecond +
        (expired ? -_validitySeconds : _validitySeconds);
    return <String>[
      _header,
      '{"exp":$exp}',
      _signature,
    ].map(_base64Url).join('.');
  }

  /// What `badRefresh` hands back, so the failure path has something unusable to reject.
  static String malformed() => 'sample-not-a-jwt';

  /// What a simulated scan links: the same unsigned shape over the chosen span and bindings, with nonsense PII.
  static String session({
    required UseSmileIDSampleSimulatedSpan span,
    required UseSmileIDSampleSimulatedBindings bindings,
    required UseSmileIDSampleEnvironment environment,
    required int nowMillis,
  }) {
    final int nowSeconds = nowMillis ~/ _millisPerSecond;
    // An ended span is minted wholly in the past, which is the only way to reach the expiry gate.
    final int issuedAt = span.inPast
        ? nowSeconds - span.span.inSeconds - _endedLagSeconds
        : nowSeconds;
    final List<String> claims = <String>[
      '"iat":$issuedAt',
      '"exp":${issuedAt + span.span.inSeconds}',
      // With the path a real claim carries, so the fixture exercises the host match.
      '"api_url":"${environment.baseUrl}$_apiPath"',
      if (bindings.binds) _payloadClaim(bindings, issuedAt),
    ];
    return <String>[
      _header,
      '{${claims.join(',')}}',
      _signature,
    ].map(_base64Url).join('.');
  }

  static String _payloadClaim(
    UseSmileIDSampleSimulatedBindings bindings,
    int issuedAtSeconds,
  ) {
    final List<String> fields = <String>[
      if (bindings.userDetails) ...<String>[
        for (final String field in _vaultedFields) '"$field":"vault_$field"',
        // The two the Portal leaves in plaintext, so a decode can read them back.
        '"country":"${UseSmileIDSampleCountry.ke.code}"',
        '"id_type":"${UseSmileIDSampleIdType.nationalId.id}"',
      ],
      if (bindings.consent) _consentClaim(issuedAtSeconds),
    ];
    return '"payload":{${fields.join(',')}}';
  }

  /// All four subfields: the SDK treats a partial binding as a build error, not a partial relaxation.
  static String _consentClaim(int issuedAtSeconds) {
    final String grantedAt =
        '${DateTime.fromMillisecondsSinceEpoch(issuedAtSeconds * _millisPerSecond, isUtc: true).toIso8601String().split('.').first}Z';
    return '"consent":{"granted":true,"granted_at":"$grantedAt",'
        '"notice_language":"en","notice_privacy_policy_url":"$_privacyPolicyUrl"}';
  }

  static String _base64Url(String value) =>
      base64Url.encode(utf8.encode(value)).replaceAll('=', '');

  static const String _header = '{"alg":"none","typ":"JWT"}';
  static const String _signature = 'sample-signature';
  static const int _millisPerSecond = 1000;
  static const int _validitySeconds = 3600;
  static const int _endedLagSeconds = 60;
  static const String _apiPath = 'v3';
  static const String _privacyPolicyUrl = 'https://smile.id/privacy-policy';
  static const List<String> _vaultedFields = <String>[
    'given_names',
    'last_name',
    'email',
    'phone_number',
    'id_number',
  ];
}
